R/paginate.R

```diff
@@ -0,0 +1,229 @@
+#' @param n_pages A function that extracts the next token from the [response].
+#'
+#' @return A modified HTTP [request].
+#' @export
+#'
+#' @examples
+#' request("https://pokeapi.co/api/v2/pokemon") %>%
+#'   req_url_query(limit = 150) %>%
+#'   req_paginate_next_url(
+#'     next_url = function(resp) resp_body_json(resp)[["next"]],
+#'     n_pages = function(resp) {
+#'       calculate_n_pages(
+#'         page_size = 150,
+#'         total = resp_body_json(resp)$count
+#'       )
+#'     }
+#'   )
+req_paginate <- function(req,
+                         next_request,
+                         n_pages = NULL) {
+  check_request(req)
+  check_function2(next_request, args = c("req", "resp"))
+  check_function2(n_pages, args = "resp", allow_null = TRUE)
+
+  req_policies(
+    req,
+    paginate = list(
+      next_request = next_request,
+      n_pages = n_pages
+    )
+  )
+}
+
+#' Perform a paginated request
+#'
+#' @inheritParams req_perform
+#' @param resp An HTTP [response].
+#' @param max_pages The maximum number of pages to request.
+#' @param progress Display a progress bar?
+#'
+#' @return A list of responses.
+#' @export
+#'
+#' @examples
+#' req_pokemon <- request("https://pokeapi.co/api/v2/pokemon") %>%
+#'   req_url_query(limit = 150) %>%
+#'   req_paginate_next_url(
+#'     next_url = function(resp) resp_body_json(resp)[["next"]],
+#'     n_pages = function(resp) {
+#'       calculate_n_pages(
+#'         page_size = 150,
+#'         total = resp_body_json(resp)$count
+#'       )
+#'     }
+#'   )
+#'
+#' responses <- paginate_perform(req_pokemon)
+paginate_perform <- function(req,
+                             max_pages = 20L,
+                             progress = TRUE) {
+  check_request(req)
+  check_bool(progress)
+
+  resp <- req_perform(req)
+  f_n_pages <- req$policies$paginate$n_pages %||% function(resp) Inf
+  n_pages <- min(f_n_pages(resp), max_pages)
+
+  out <- vector("list", length = n_pages)
+  out[[1]] <- resp
+
+  cli::cli_progress_bar(
+    "Paginate",
+    total = n_pages,
+    format = "{cli::pb_spin} Page {cli::pb_current}/{cli::pb_total} | ETA: {cli::pb_eta}",
+    current = 1L
+  )
+
+  for (page in seq2(2, n_pages)) {
+    req <- paginate_next_request(resp, req)
+    if (is.null(req)) {
+      page <- page - 1L
+      break
+    }
+
+    resp <- req_perform(req)
+
+    body_parsed <- resp_body_json(resp)
+    out[[page]] <- resp
+
+    cli::cli_progress_update()
+  }
+  cli::cli_progress_done()
+
+  # remove unused end of `out` in case the pagination loop exits before all
+  # `max_pages` is reached
+  if (page < n_pages) {
+    out <- out[seq2(1, page)]
+  }
+
+  out
+}
+
+#' @rdname paginate_perform
+paginate_next_request <- function(resp, req) {
+  check_response(resp)
+  check_request(req)
+
+  if (!req_policy_exists(req, "paginate")) {
+    cli::cli_abort(c(
+      "{.arg req} doesn't have a pagination policy",
+      i = "You can add pagination via `req_paginate()`."
+    ))
+  }
```

Comment:
Maybe move this check to `paginate_perform`?

Reply:
I thought it makes sense to also export `paginate_next_request()` so I left the check there but also added one to `paginate_req_perform()`.